---
title: "FigureYa41GEO2lncRNA_update"
params:
  author: "Ying Ge, Yijing Chen"
output: html_document
---

**Author(s)**: `r params$author`  
**Date**: `r Sys.Date()`  


# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述
## Requirement description

从GEO下载芯片数据，从基因表达矩阵分别提取lncRNA、mRNA、miRNA的表达矩阵

这套代码使用三个相对独立的模块来解决上述需求，可根据自己的实际需要，灵活使用这三个模块：

【模块一】从GEO下载microarray数据，不包括测序数据。

【模块二】根据Ensembl的BioMart对基因的biotype注释，识别出哪些基因是lncRNA、哪些是mRNA、哪些是miRNA。

不仅限于芯片数据，只要有gene symbol或Ensembl ID就能识别。

如果没有gene symbol或Ensembl ID，或其他任何跟BioMart共有的ID，就需要做序列比对，本套代码不包含序列比对功能。

【模块三】根据基因名提取对应的表达矩阵。

不仅限于GEO数据，还适用于从TCGA下载的数据、你自己的测序数据。只要提供一个基因列表和一个表达矩阵文件。

**注：**如果想要miRNA成熟体的表达矩阵，最好的方式去找miRNA芯片数据，例如GSE113596，用【模块一】下载即可。如果用【模块二】和【模块三】从普通芯片数据里提取miRNA，那其实是miRNA前体的杂交信号。

Download the microarray data from GEO, and extract the expression matrix of lncRNA, mRNA and miRNA from the gene expression matrix respectively.

This set of code uses three relatively independent modules to solve the above needs, you can use these three modules flexibly according to your actual needs:

[Module I] Download microarray data from GEO, excluding sequencing data.

[Module 2] Identify which genes are lncRNAs, which are mRNAs, and which are miRNAs based on the biotype annotations of the genes by Ensembl's BioMart.

This is not limited to microarray data, as long as a gene symbol or Ensembl ID is available, identification is possible.

If there is no gene symbol or Ensembl ID, or any other ID shared with BioMart, you need to do sequence comparison, and this code does not contain sequence comparison function.

[Module 3] Extract the corresponding expression matrix according to the gene name.

Not only limited to GEO data, but also applicable to data downloaded from TCGA, your own sequencing data. Just provide a gene list and an expression matrix file.

**Note:** If you want the expression matrix of miRNA mature body, the best way is to find the miRNA microarray data, for example, GSE113596, and download it with [Module 1]. If you use [Module 2] and [Module 3] to extract miRNA from normal microarray data, it is actually hybridization signals of miRNA precursors.

## 应用场景
## Application scenario

场景一：想计算lncRNA、miRNA、mRNA间的相关性，用来找miRNA、lncRNA的靶基因，或者用功能已知的mRNA推测lncRNA的功能，首先就要获得lncRNA、miRNA跟mRNA的表达矩阵。需要从GEO下载芯片数据，就从【模块一】开始；以TCGA表达矩阵作为输入（FigureYa22、23、34的输出文件），就从【模块二】开始。

场景二：打算做RNA-seq，正在设计实验，不知道该做哪种处理、选什么时间点。那就先看看别人的实验设计获得的数据效果如何，用【模块一】下载GEO的芯片数据。

场景三：TCGA里感兴趣的癌症类型样本量太少或没有，或者想拿多个来源的数据验证，就用【模块一】下载GEO的芯片数据。

> 这篇为了兼顾特殊情况，文字越写越多。如果你的数据属于大多数，就不用看那么多文字，只把GSE和GPL或者“easy_input.csv”文件替换成你自己的数据，运行代码就好。

Scenario 1: To calculate the correlation among lncRNA, miRNA, and mRNA for the purpose of identifying target genes of miRNA and lncRNA, or to infer the function of lncRNA based on functionally known mRNA, it is first necessary to obtain the expression matrices of lncRNA, miRNA, and mRNA. If you need to download microarray data from GEO, start from [Module 1]; if you are using the TCGA expression matrix as input (output files from FigureYa22, 23, and 34), start from [Module 2].

Scenario 2: You are planning to do RNA-seq, and are designing the experiment, and you don't know which treatment to do and what time point to choose. Then see how the data obtained from other people's experimental design works first, and use [Module 1] to download the  microarray data from GEO.

Scenario 3: If the cancer type of interest in TCGA has too few samples or is not available, or if you want to validate using data from multiple sources, use [Module 1] to download microarray data from GEO.

> This article includes more text to accommodate special cases. If your data belongs to the majority, you don’t need to read all the text, just replace the GSE and GPL or the “easy_input.csv” file with your own data and run the code.  

## 环境设置
## Environment setting



```r
#use the domestic mirror installation package
BiocManager::install("GEOquery")
BiocManager::install("biomaRt")
```


load package

```{r}
source("install_dependencies.R")
library(Biobase)
library(GEOquery)
library(biomaRt)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 display English error messages
options(stringsAsFactors = FALSE) #禁止chr转成factor prohibit the conversion of chr to factor
```

## 【模块一】从GEO下载基于microarray的基因表达矩阵。
## [Module 1] Download microarray-based gene expression matrix from GEO.

如果你用FigureYa22、23、34获得了TCGA的基因表达矩阵，已经保存为`easy_input.csv`的格式，不需要从GEO下载表达矩阵，就跳过这步，直接进入【模块二】。

- 第一步：到NCBI的GEO数据库查询你感兴趣的数据，<https://www.ncbi.nlm.nih.gov/geo/>

- 第二步：借助GEOquery，提取表达数据。

**注意：**把GSE替换成你想用的数据，切记同时替换成这套数据所用的GPL（哪个芯片平台），才能获得正确的探针组注释信息。**GPL去哪里找？**在<https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE36895>页面搜platform，它右边就是GPL。

If you have used FigureYa22, 23, 34 to obtain the gene expression matrix of TCGA, which has been saved in the format of `easy_input.csv`, and don't need to download the expression matrix from GEO, skip this step and go directly to [Module 2].

- Step 1: Go to NCBI's GEO database to query the data you are interested in, <https://www.ncbi.nlm.nih.gov/geo/>

- Step 2: With the help of GEOquery, extract the expression data.

**Note:** Replace GSE with the data you want to use, and remember to also replace it with GPL (which chip platform) used for this set of data in order to get the correct probe set annotation information. **Where to find the GPL? **Search platform on the <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE36895> page, and the GPL is listed to the right of it.

```{r,message=FALSE}
gset <- getGEO("GSE36895", GSEMatrix =TRUE, getGPL = TRUE, AnnotGPL = TRUE) 
#如果提示没有GPL***.anno.gz文件，就用下面这句
#If prompted that there is no GPL***.anno.gz file, use the following sentence
#gset <- getGEO("GSE36895", GSEMatrix =TRUE, getGPL = TRUE) 

if (length(gset) > 1) idx <- grep("GPL570", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]

#查看gset里的丰富信息
#view the wealth of information in gset
#str(gset)
#如果gset里没有gene symbol或其他任何跟ensembl共有的ID，就需要用序列比对的方式去找这段探针组所对应的基因名，才能提取lncRNA。
#if there is no gene symbol or any other ID shared with ensembl in the gset, you need to find the gene name corresponding to this probe set by sequence comparison to extract lncRNA.

#提取表达矩阵
#extract the expression matrix
exprdf<-data.frame(exprs(gset))
dim(exprdf)

#默认样品名是geo_accession，例如GSM904985
#还可以提取title作为样品名，例如Normal cortex of patient 14
#default sample name is geo_accession, e.g. GSM904985
#title can also be extracted as sample name, e.g. Normal cortex of patient 14
#colnames(exprdf)<-gset@phenoData@data$title

#保存到文件
#save to file
#write.csv(exprdf,"not_easy_input.csv",quote = F)
```

现在你就获得了这套芯片的表达矩阵

这时，行名是探针组的名字，不好用，我们需要给它加上基因名。

你可能会遇到三种情况：

1. 像示例所用的芯片注释文件提供了gene symbol，我们就提取gene symbol。

2. 如果你用的芯片连gene symbol都没有，就去gset和【模块二】的listAttributes里找它俩共有的ID，然后替换下面的Gene symbol。

3. 如果没有任何共有ID，就需要你自己通过序列比对去给芯片做注释，找到序列对应的基因名，【模块一】对于你来说，到这里就暂停了。

Now you get the expression matrix of this chip set

At this point, the row name is the name of the probe set, which is not good, and we need to add gene names to it.

You may encounter three situations:

1. If the chip annotation file used in the example provides gene symbol, we extract the gene symbols.

2. If the chip you use doesn't even have a gene symbol, go to the listAttributes of gset and [Module 2] to find the ID shared by both of them, and then replace the gene symbol below.

3. If there is no shared ID, you need to annotate the chip by sequence comparison to find the gene name corresponding to the sequence, [Module 1] is suspended here for you.

```{r}
exprdf$gsym<-gset@featureData@data$`Gene symbol`
#有时要用下面这行
#sometimes you have to use the following line
#exprdf$gsym<-gset@featureData@data$GENE_SYMBOL

#删除没有gene symbol的探针组
#delete probe sets without gene symbols
exprdf<-exprdf[exprdf$gsym!="",]
dim(exprdf)

#有的探针组对应多个基因，用“///”分隔基因名，删掉这样的行
#there are probe sets corresponding to multiple genes, use "///" to separate gene names, delete such lines
exprdf<-exprdf[!grepl("///", exprdf$gsym),]
dim(exprdf)

#有多个探针组对应同一个基因，取中值
#如果想取平均值，就把median改为mean
#if there is more than one probe set corresponding to the same gene, take the median value
#if you want to take the mean, change median to mean
exprdf_uniq<-aggregate(.~gsym,exprdf,median)
dim(exprdf_uniq)

#现在就可以用gene symbol作为行名了
#now it's possible to use the gene symbol as row name
rownames(exprdf_uniq)<-exprdf_uniq$gsym
#删除gene symbol列
#delete the gene symbol column
exprdf_uniq<-subset(exprdf_uniq,select = -gsym)

#保存所有基因的表达矩阵到文件
#save the expression matrix of all genes to a file
#write.csv(exprdf_uniq, file = "gene_exp.csv", row.names = T, quote = F)
#此处仅保存前4个sample
#only the first 4 samples are saved here
write.csv(exprdf_uniq[,1:4],file="easy_input.csv",row.names = T,quote = F)
```

**题外话：**以“GSE36895”为例，在<https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE36895>页面底部有个“Analyze with GEO2R”按钮，点击按钮，就进入了GEO2R界面，<https://www.ncbi.nlm.nih.gov/geo/geo2r/>，点击Define groups，输入分组信息，然后就能做**简单的差异基因筛选和画图**。其实都是用R实现的，点击选项卡里的R script，就能看到R代码。

**Off-topic:** Take “GSE36895” as an example, at the bottom of the <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE36895> page, there is a button "Analyze with GEO2R", click the button to enter the GEO2R interface, <https://www.ncbi.nlm.nih.gov/geo/geo2r/>, click Define groups, input the group information, and then you can do **simple differential gene screening and drawing**. Actually it is all realized with R, click R script in the tab, you can see the R code.

## 【模块二】用Ensembl的biomaRt，识别出哪些基因是lncRNA、哪些是mRNA、哪些是miRNA。
## [Module 2] Use Ensembl's biomaRt to identify which genes are lncRNAs, which are mRNAs, and which are miRNAs.

### 输入文件
### Input file

用【模块一】获得的`easy_input.csv`作为输入，基因名是gene symbol；

或者FigureYa22、23、34的输出文件作为输入，基因名是ensembl ID；

或者你自己写的基因列表，至少包含第一列：基因名，可以是gene symbol，或者ensembl ID。

第二列开始是表达量，非必须。

Use `easy_input.csv` obtained from [Module 1] as input, with gene name as gene symbol;

Or the output files from FigureYa22, 23, 34 as input, with gene name as ensembl ID;

Or your own list of genes, written by yourself, containing at least the first column: gene name, which can be either gene symbol, or ensembl ID.

The second column starts with the expression, not required.

```{r}
exprdf_uniq <- read.csv("easy_input.csv",header = T,row.names = 1)
rownames(exprdf_uniq)[1:4]
```

### 从biomaRt提取gene biotype
### Extract gene biotype from biomaRt

这里用到biomaRt包，来自ensembl。

- 第一步，选择你要用的基因组版本。

此处用人类ensembl最新版本，如果想用旧的基因组版本或其他物种，需要按照注释修改host = 后面的参数，

点击链接，<http://asia.ensembl.org/info/website/archives/assembly.html>，查看genome assembly跟enssembl版本的对应关系

The biomaRt package is used here, from ensembl.

- Step 1: select the version of the genome you want to use.

Here we use the latest version of human ensembl, if you want to use an older genome version or another species, you need to change the parameter after host = according to the comment.

Click on the link, <http://asia.ensembl.org/info/website/archives/assembly.html>, to see the correspondence between genome assembly and enssembl version

```{r}
#用下面这行查看ensembl基因组版本跟host的对应关系
#use the following line to see the correspondence between the Ensembl genome version and the host
#listEnsemblArchives()

#运行下面两行，查看基因组
#run the following two lines to view the genome
#mart = useMart('ensembl')
#listDatasets(mart)
#你需要哪个物种，就复制它在dataset列里的词，放在下面这行的`dataset = `参数里
#which species you need, copy its word in the dataset column and put it in the `dataset = ` parameter on the following line
ensembl <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", 
                   dataset = "hsapiens_gene_ensembl", #人 human
                   #dataset = "mmusculus_gene_ensembl", #小鼠 mouse
                   #dataset = "rnorvegicus_gene_ensembl", #大鼠 rat
                   #dataset = "dmelanogaster_gene_ensembl", #果蝇 fruit fly
                   host = "www.ensembl.org") 
```

```r
#植物用下面三行，以拟南芥为例
#use the following three lines for plants, using Arabidopsis as an example
mart <- useMart(biomart = 'plants_mart', host = "plants.ensembl.org")
listDatasets(mart)
ensembl <- useMart(biomart = "plants_mart",
                   dataset = "athaliana_eg_gene", 
                   host = "plants.ensembl.org") 
```

- 第二步，提取biotype

我们需要用gene symbol或ensembl ID来提取gene_biotype

然后用gene biotype区分lncRNA、miRNA和mRNA

- Step 2: Extract the biotype

We need to extract gene_biotype with gene symbol or ensembl ID

Then use the gene biotype to distinguish lncRNA, miRNA and mRNA

```{r,message=FALSE}
#查看Filters和Attributes提供了哪些信息
#会看到基因名gene symbol在第62行
#你的芯片数据不一定在第62行哦，要具体看的
#see what information is provided by Filters and Attributes
#you will see the gene symbol in line 62
#your microarray data may not be on line 62, you need to check it specifically
#listFilters(ensembl)
#listAttributes(ensembl)
listAttributes(ensembl)[62,]
#下面的“filters = ”要用到这个name
#the following "filters = " will use this name

feature_info <- getBM(attributes = c("gene_biotype",
                                     #"transcript_biotype",#还可以提取transcript_biotype transcript_biotype can also be extracted
                                     #如果基因名是gene symbol，就运行下面这行
                                     #if the gene name is gene symbol, run the following line
                                     "hgnc_symbol"), 
                                     #如果基因名是ensembl ID，就运行下面这行 if the gene name is ensembl ID, run the following line
                                     #"ensembl_gene_id"),
                      #如果基因名是gene symbol，就运行下面这行
                      #if the gene name is gene symbol, run the following line
                      filters = "hgnc_symbol", #小鼠是mgi_symbol，大鼠是mgi_symbol mouse is mgi_symbol, rat is mgi_symbol
                      #如果基因名是ensembl ID，就运行下面这行
                      #if the gene name is ensembl ID, run the following line
                      #filters = "ensembl_gene_id",
                      values = rownames(exprdf_uniq), mart = ensembl)

#有些芯片注释的gene symbol跟最新版本ensembl的基因名不一致，需要返回上一步，换比较老的版本。
#TCGA数据的ensembl ID跟最新版ensembl一致
#the gene symbol of some chip annotations is inconsistent with the gene name of the latest version of ensembl, you need to return to the previous step and change to the older version.
#the ensembl ID of TCGA data is consistent with the latest version of ensembl
if (nrow(exprdf_uniq) != nrow(feature_info)){
  #查看哪些基因名不一致
  #view which gene names are inconsistent
  library(dplyr)
  diffName<-setdiff(rownames(exprdf_uniq),feature_info[,2])
  length(diffName)
  head(diffName)
}

length(unique(feature_info$hgnc_symbol))
#有些gene symbol对应多个ensembl id，因此会有多个biotype，例如
#some gene symbols correspond to multiple ensembl ids, resulting in multiple biotypes, for example
feature_info[feature_info$hgnc_symbol == "ARHGEF26-AS1",]
#TCGA数据不会遇到这个问题，因为ensembl id跟gene_biotype是一一对应的关系
#TCGA data does not encounter this problem because ensembl id and gene_biotype have a one-to-one correspondence

#把基因的biotype保存到文件
#save the biotype of the gene to a file
write.csv(feature_info[,c(2,1)],"gene_biotype.csv",quote = F,row.names = F)
```

### 识别lncRNA、mRNA和miRNA
### Identify lncRNA, mRNA, and miRNA

对lncRNA的定义，可参考Vega的标准：
<http://vega.archive.ensembl.org/info/about/gene_and_transcript_types.html>

The definition of lncRNA can refer to the standard of Vega: <http://vega.archive.ensembl.org/info/about/gene_and_transcript_types.html>

```{r}
#查看gene biotype的类型
#view the type of gene biotype
unique(feature_info$gene_biotype)

#此处定义protein_coding作为mRNA
#protein_coding is defined here as the mRNA
mRNA <-"protein_coding"
#定义lncRNA
#define lncRNA
lncRNA <- "lncRNA"
#还可以定义miRNA
#miRNA can also be defined
miRNA <-"miRNA"

#下面就是表达矩阵里的mRNA、lncRNA、miRNA及其数量，把每类基因的基因名保存到相应的文件里。
#Below are the mRNA, lncRNA, and miRNA in the expression matrix along with their quantities, save the gene names of each gene category into the corresponding file.
mRNA.list<-feature_info[grepl(mRNA, feature_info$gene_biotype),]
write.table(mRNA.list,"mRNA.list.txt",quote = F,row.names = F, col.names = F)
nrow(mRNA.list)

lncRNA.list<-feature_info[grepl(lncRNA, feature_info$gene_biotype),]
write.table(lncRNA.list,"lncRNA.list.txt",quote = F,row.names = F, col.names = F)
nrow(lncRNA.list)

miRNA.list<-feature_info[grepl(miRNA, feature_info$gene_biotype),]
write.table(miRNA.list,"miRNA.list.txt",quote = F,row.names = F, col.names = F)
nrow(miRNA.list)
```

## 【模块三】根据基因名提取对应的表达矩阵
## [Module 3] Extract the corresponding expression matrix based on gene name

### 输入文件
### Input file

两个输入文件：

- 表达矩阵：easy_input.csv。第一列：基因名。可以是gene symbol（【模块一】的输出文件），或者ensembl_gene_id（FigureYa22、23、34输出文件）。第二列开始是表达量。每行一个基因，每列一个sample。

- 基因列表：mRNA.list.txt。【模块二】的输出文件。包含一列基因名，可以是gene symbol，或者ensembl_gene_id，要跟表达矩阵的基因名一致。

Two input files:

- Expression matrix: easy_input.csv. first column: gene name. It can be gene symbol ([Module 1] output file), or ensembl_gene_id (FigureYa22, 23, 34 output files). The second column starts with the expression. One gene per row, one sample per column.

- Gene list: mRNA.list.txt. output file of [Module 2]. It contains a column of gene names, either gene symbol, or ensembl_gene_id, and must be consistent with the gene names in the expression matrix.  

```{r}
exprdf_uniq<-read.csv("easy_input.csv",header = T,row.names = 1)
head(exprdf_uniq)

#以lncRNA基因列表为例
#take the lncRNA gene list as an example
lncRNA.list<-read.table("lncRNA.list.txt")
head(lncRNA.list)
```

### 提取表达矩阵，保存到文件
### Extract the expression matrix and save it to a file

```{r}
mRNA_expr <- exprdf_uniq[as.character(mRNA.list[,2]),]
write.csv(mRNA_expr,"mRNA_expr.csv",quote = F,row.names = T)
head(mRNA_expr)

lncRNA_expr <- exprdf_uniq[as.character(lncRNA.list[,2]),]
write.csv(lncRNA_expr,"lncRNA_expr.csv",quote = F,row.names = T)
head(lncRNA_expr)

miRNA_expr <- exprdf_uniq[as.character(miRNA.list[,2]),]
write.csv(miRNA_expr,"miRNA_expr.csv",quote = F,row.names = T)
head(miRNA_expr)
```

```{r}
sessionInfo()
```