---
title: "FigureYa23count2TPMv2"
params:
  author: "Zhougeng Xu; Ying Ge, Yijing Chen"
output: html_document
---

**Author(s)**: `r params$author`  
**Date**: `r Sys.Date()`  


![](hoptop.png)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述
## Requirement description

RNA-seq read count转换成TPM。

Convert RNA-seq read count to TPM.

## 应用场景
## Application scenario

只能拿到RNA-seq数据的read count，想转换成TPM。

注：TPM的计算方式最好是直接从原始的FASTQ文件进行获取，例如Samlon, Sailfish 或 kallisto, 或者先用FASTQ比对得到BAM，然后用RESM估计TPM.

If you only have access to RNA-seq read counts and want to convert them to TPM. 

Note: The best way to calculate TPM is directly from the original FASTQ files using tools like Salmon, Sailfish, or kallisto, or by first aligning the FASTQ to obtain BAM files and then using RESM to estimate TPM. 

## 计算基因长度
## Calculate gene length

这步比较耗时。建议：

- 使用TCGA数据的小伙伴可以直接使用压缩包里的`eff_length.csv`文件，就可以跳过这步，直接从“输入数据”开始运行。

- 每个物种每个基因组注释版本的基因长度是相同的，下面这部分代码你只需要运行一次，保存好生成的`eff_length.csv`文件，标注好基因组版本。当你需要做count到FPKM或TPM的转换时，就可以跳过“计算基因长度”这步，直接从“read count转TPM”开始运行。

这一步有两种思路:

- 方法1：简单粗暴的计算基因在染色体的起始和结束之差
- 方法2：比较麻烦的计算每个基因的最长转录本(外显子之和)或所有外显子之和

推荐方法2

This step is more time consuming. It is recommended:

- If you use TCGA data, you can directly use the `eff_length.csv` file in the zip package, then you can skip this step and run directly from "Input data".

- The gene length is the same for each genome annotation version of each species, you only need to run this part of the code once, save the generated `eff_length.csv` file, and label the genome version. When you need to do the conversion from count to FPKM or TPM, you can skip the step of "calculate gene length", and directly start from "read count to TPM".

There are two ways of thinking in this step:

- Method 1: Simply calculate the difference between the start and end of the gene on the chromosome.

- Method 2: Calculate the longest transcript (sum of exons) of each gene or the sum of all exons, which is more troublesome.

Recommended method 2

### 方法1
### Method 1

如果只是想粗略了解一下表达情况，可以简单把基因在染色体上的起始位置和结束位置之差用作标准化的长度。

If you just want to get a rough idea of the expression, you can simply use the difference between the start and end positions of the gene on the chromosome as a normalized length.

```r
library(biomaRt)

#查看基因组参数
#view genomic parameter
mart = useMart('ensembl')
listDatasets(mart)

#你需要哪个基因组，就复制它在dataset列里的词，放在下面这行的`dataset = `参数里
#此处以人类为例，植物参考注一
#Which genome you need, copy its word in the dataset column and put it in the `dataset = ` parameter on the following line
#Here is an example for humans, for plants refer to note 1
bmart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL", 
                          dataset = "hsapiens_gene_ensembl",
                          host = "www.ensembl.org")

# 从输入数据里提取基因名
# extract gene names from input data
feature_ids <- rownames(expMatrix)

attributes = c(
        "ensembl_gene_id",
        #"hgnc_symbol",
        "chromosome_name",
        "start_position",
        "end_position"
    )
filters = "ensembl_gene_id"

feature_info <- biomaRt::getBM(attributes = attributes, 
                               filters = filters, 
                               values = feature_ids, mart = bmart)
mm <- match(feature_ids, feature_info[[filters]])
feature_info_full <- feature_info[mm, ]
rownames(feature_info_full) <- feature_ids

# 计算基因的有效长度
# calculate the effective length of the gene
eff_length <- abs(feature_info_full$end_position - feature_info_full$start_position)
names(eff_length) <- feature_info_full$ensembl_gene_id
```

这里的`eff_length`就是每个基因的有效长度

Here `eff_length` is the effective length of each gene

### 方法2
### Method 2

这一步原本打算用Bioconductor的`ord.Hs.eg.db`,但是实际发现里面的ENSEMBL ID明显比输入的TCGA数据集里的ID少。所以决定从GTF文件提取：

先写一个函数，实现从下载GTF文件到计算基因长度.

分为两个模式，用模式2算出的length去做count转FPKM，跟TCGA的FPKM最为接近，详情请看FigureYa34count2FPKM。自己斟酌选用：

- 模式1. transcript，计算最长的转录本长度

- 模式2. gene，计算所有转录本的所有外显子的总长度

This step was initially intended to use Bioconductor's `org.Hs.eg.db`, but it was actually found that the ENSEMBL IDs in it were significantly less than those in the input TCGA dataset. Therefore, it was decided to extract them from the GTF file:

First write a function to realize from download GTF file to calculate gene length.

There are two modes, use the length calculated in mode 2 to do the count to FPKM, which is the closest to the FPKM of TCGA, please see FigureYa34count2FPKM for details. choose at your own discretion:

- Model 1. transcript, calculate the length of the longest transcripts.

- Model 2. gene, calculate the total length of all exons of all transcripts.

```r
##' download or load local gtf file and get the longest transcript of each gene
##'
##' @title get_eff_len
##' @param file url or local file path of gff3 file
##' @param mode = transcript or gene
##' @param feature CDS or exon
##' @param gene_id the feature name of gene id in gtf
##' @param transcript_id the feature name of transcript id in gtf
##' @export
##' @author XuZhougeng
get_eff_len <- function(file=NULL, 
                        mode='gene',
                        feature = "exon",
                        gene_id = 'gene_id',
                        transcript_id = "transcript_id", ...){
  # 检查是否提供文件参数
  # check if file parameter is provided
  if ( is.null(file)){
    stop(" no file path or url provided")
  }
  
  file_split <- unlist(strsplit(file,'/'))
  file_name <- file.path(Sys.getenv('R_USER'),
                         file_split[length(file_split)])
  
  # 检查文件是 url 还是本地路径
  # check file is url or local path
  if (grepl(pattern = '(ftp|http|https)', file)){
    file_name <- file_name
    if ( ! file.exists(file_name)){
      print(paste0("file will save in ", file_name))
      download.file(url = file, 
                    destfile = ,file_name,
                    method = "internal")
        }
  } else{
    if ( ! file.exists(file)){
      stop("file is not exists")
    }
    file_name <- file
  }

  # 从文件名获取文件类型
  # 用不同的读取函数建立连接
  # get the file type from the file name
  # and build connection  with different read function
  file_type <- sub('.*?\\.(.*?)$','\\1',file_name)
  if (file_type == 'gtf'){
    gtf_in <- file(file_name, open='rt')
  } else if(file_type == 'gz'){
    file_name_tmp <- gsub(sprintf("[.]%s$",'gz'),"",file_name)
    if ( ! file.exists(file_name_tmp)) 
      R.utils::gunzip(file_name, remove = FALSE)
    gtf_in <- file(file_name_tmp, open='rt')
  } else{
    err_message <- paste0("unsupport file type: ", file_name)
    stop(err_message)
  }

  gtf_lines <- readLines(gtf_in, warn=FALSE)
  close(gtf_in)
  
  # 保存基因的有效长度
  # save the effitve length of gene
  feature_pattern <- paste0("\\t",feature,"\\t")
  mt_rows <- sum(grepl(feature_pattern, gtf_lines))
  # 检查feature
  # check the feature
  if ( mt_rows <= 1){
    features <- do.call(rbind, strsplit(gtf_in, '\\t'))[,3]
    print("feature is unfoundable, the top 5 feature is:")
    print(as.data.frame(table(feature))[1:5,])
    stop(paste0("set your feature and run again"))
  }
  
  mt <- matrix(nrow = mt_rows, ncol= 2)
  factor_names <- vector(mode = "character",length = mt_rows)
  # 保存上一行的基因 ID 
  # 迭代 
  # save the gene id of the previous line 
  # iteration 
  line_num <- 1
  for (line in gtf_lines){
    if ( grepl(pattern = "^#", x = line)) next
    records <- unlist(strsplit(x = line, '\\t'))
    if ( grepl(feature, records[3]) ){
      gene_pattern <- paste0(".*?", gene_id, " \"(.*?)\";.*")
      gene_id_ <- sub(gene_pattern,"\\1",records[9])
      tx_pattern <- paste0(".*?", transcript_id, " \"(.*?)\";.*")
      tx_id_ <- sub(tx_pattern, "\\1",records[9])
      gene_start <- as.numeric(records[4])
      gene_end <- as.numeric(records[5])
      mt[line_num,] <- c(gene_start, gene_end)
      # 使用"___"而不是"_"作为分割符，以避免 
      # 基因 ID 中有"_"的问题
      # using "___" instead of "_" as spliter to avoid 
      # problem that "_" is in gene id
      factor_names[line_num] <- paste0(gene_id_,"___",tx_id_)
      line_num <- line_num + 1
    }
  }
  
  if ( mode == 'gene'){
    # 计算最长的转录本
    # calculate the longest transcript
    by_gene <- as.factor(do.call(rbind, strsplit(factor_names,'___'))[,1])
    gene_length <- sapply(split(mt[,c(1,2)], by_gene), my.reduce)
  }else if(mode=='transcript'){
    # 计算单个基因中的所有外显子
    # calculate all exons in a single gene
    by_transcript <- as.factor(factor_names)
    tx_length <- sapply(split(mt[,c(1,2)], by_transcript), my.reduce)
    by_gene <- as.factor(do.call(rbind, 
                                  strsplit(names(tx_length), '___'))[,1])
    gene_length <- sapply(split(tx_length, by_gene), max)
  }else{
    stop("unsupport mode")
  }
  
  return(gene_length)
}

my.reduce <- function(x){
  tx_num <- length(x) / 2
  total <- c()
  for (i in seq(1,tx_num)){
    total <- c(seq(x[i], x[i+tx_num]), total)
  }
  length(unique(total))
}


```

> 目前只测试了拟南芥和人类。
> 注，函数不支持需要登陆才能下载的GTF的URL

下面调用函数下载GTF文件，提取存放有效转录本长度的向量

把你想要的gtf文件的地址写在下面`url <- `的后面

案例一：

此处以TCGA的read count作为输入，用跟TCGA一致的注释文件提取外显子长度。

到这里查询TCGA用的是哪个版本的注释文件：<https://gdc.cancer.gov/about-data/data-harmonization-and-generation/gdc-reference-files>

用的是这个版本：`GDC.h38 GENCODE v22 GTF (used in RNA-Seq alignment and by HTSeq)`

去<https://www.gencodegenes.org/releases/22.html>下载相应版本的gtf文件

> Currently only Arabidopsis thaliana and human are tested.
> Note, the function does not support the URL of GTF which requires login to download.

Call the function below to download the GTF file and extract the vector containing the lengths of valid transcripts.

Write the address of the GTF file you want in the following `url <- ` followed by

Case 1:

Here, the read count of TCGA is used as input, and the exon lengths are extracted using an annotation file consistent with TCGA.

Go here to query which version of the annotation file is used by TCGA: <https://gdc.cancer.gov/about-data/data-harmonization-and-generation/gdc-reference-files>

This version is used: `GDC.h38 GENCODE v22 GTF (used in RNA-Seq alignment and by HTSeq)`

Go to <https://www.gencodegenes.org/releases/22.html> to download the corresponding version of the gtf file

```r
#url <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_22/gencode.v22.annotation.gtf.gz"

file<- "gencode.v22.annotation.gtf"
eff_length <- get_eff_len(file = file,
                          #采用模式2，计算所有转录本的所有外显子的总长度
                          #Use model 2 to calculate the total length of all exons of all transcripts
                          mode='gene',
                          #采用模式1，计算最长的转录本长度
                          #Use model 1 to calculate the longest transcript length
                          #mode=='transcript',
                          feature = "exon",
                          gene_id = 'gene_id',
                          transcript_id = "transcript_id")

eff_length1 <- data.frame(eff_length)
write.csv(eff_length1, "eff_length.csv", row.names = TRUE)
```

> 注意，用CDS时不会统计非编码RNA

案例二：拟南芥的GTF文件(推荐Araport)，统计每个基因最长的转录本的CDS序列长度

> Note that non-coding RNA is not counted when using CDS

Case 2: GTF file for Arabidopsis thaliana (Araport recommended), calculate the CDS sequence length of the longest transcript for each gene.

```r
file <- "Araport11_GFF3_genes_transposons.201606.gtf"
eff_length <- get_eff_len(file = file,
                          feature = "CDS",
                          gene_id = 'gene_id',
                          transcript_id = "transcript_id")
```

## 下载TCGA RNA-seq的read count数据 
## Download TCGA RNA-seq read count data

如果你自己的RNA-seq数据已经保存成`easy_input.txt`的格式，就跳过这步，直接进入“输入数据”

If your own RNA-seq data has been saved in `easy_input.txt` format, skip this step and go directly to “Input data”.

```r
#source("https://bioconductor.org/biocLite.R")
#biocLite("TCGAbiolinks")
library(TCGAbiolinks)
expquery <- GDCquery(project = "TCGA-LIHC", 
                data.category = "Transcriptome Profiling",
                data.type = "Gene Expression Quantification",
                workflow.type = "HTSeq - Counts"
                )
GDCdownload(expquery)
expquery2 <- GDCprepare(expquery)
expMatrix <- TCGAanalyze_Preprocessing(expquery2)

#运行下面这行，会把read count保存到文件里，用来做更多其他的分析。
#Running the following line will save the read count to a file that can be used for more other analysis.
#write.table(expMatrix, "easy_input.txt", sep="\t", quote=F, row.names=T)
#此处保存前4个sample
#Save the first 4 samples here
write.table(expMatrix[,1:4], "easy_input.txt", sep="\t", quote=F, row.names=T)
```

## 输入数据
## Input data

第一列是基因ID，第一行是sample ID。

一个单元格内是一个基因在一个sample中的read count。

The first column is the gene ID and the first row is the sample ID.

Inside a cell is the read count of a gene in a sample.

```{r}
expMatrix <- read.table("easy_input.txt",
                        row.names = 1, header = TRUE, sep="\t")
#查看前三个基因的read count
#view the read count for the first three genes
expMatrix[1:3,]
```

## read count转TPM
## Read count to TPM

首先要保证表达矩阵的行名和存放基因长度向量的名字一致, 这一步非常重要

It is important to make sure that the row names of the expression matrix match the names of the gene length vectors.

```{r}
eff_length2 <-read.csv("eff_length.csv", row.names = 1, header = T)
eff_length2$gene_id <- rownames(eff_length2)
rownames(eff_length2) <- do.call(rbind,strsplit(eff_length2$gene_id,'\\.'))[,1]

feature_ids <- rownames(expMatrix)

if (! all(feature_ids %in% rownames(eff_length2))){
  tbl <- table(feature_ids %in% rownames(eff_length2))
  msg1 <- sprintf("%i gene is shared, %i gene is specified", tbl[[2]],tbl[[1]])
  warning(msg1)
  
} 

if (! identical(feature_ids, rownames(eff_length2))){
  msg2 <- sprintf("Given GTF file only contain %i gene, but experssion matrix has %i gene", nrow(eff_length2), nrow(expMatrix))
  warning(msg2)
}

# 修剪表达矩阵和有效基因长度
# trim the expression matrix and effetive gene length
expMatrix <- expMatrix[feature_ids %in% rownames(eff_length2),]
mm <- match(rownames(expMatrix), rownames(eff_length2))
eff_length2 <- eff_length2[mm, ]

if (identical(rownames(eff_length2), rownames(expMatrix))){
  print("GTF and expression matix now have the same gene and gene in same order")
}

```

如果上面代码运行时有警告，主要是GTF里面的基因数低于表达矩阵时，请换一个更新版本的GTF文件. 最后结果时删减表达矩阵的行数，也就是基因数，保证表达矩阵的基因数目和GTF文件解析得到的基因数相同，并且顺序一致。

最后执行下面的代码

If there is a warning when running the above code, mainly because the number of genes in the GTF is lower than the expression matrix, please change to a newer version of GTF file. In the final result, delete the number of rows of the expression matrix, that is, the number of genes, to ensure that the number of genes in the expression matrix is the same as the number of genes in the parsed GTF file, and the order is consistent.

Finally, execute the following code

```{r}
x <- expMatrix / eff_length2$eff_length
expMatrix_tpm <- t( t(x) / colSums(x) ) * 1e6 
```

把算好的TPM保存到本地

Save the calculated TPM locally

```{r}
write.table(expMatrix_tpm, "output_count2tpm.txt", sep="\t", quote=F, row.names=T)
```

## 参考资料
## Reference

- <https://biology.stackexchange.com/questions/64860/what-is-the-length-of-gene-when-calculating-tpm-transcripts-per-million>
- `Seurat::getBMFeatureAnnos`


注1: 直接比较FPKM转成的TPM和count转换而成的TPM，完全一致。

Note 1: Direct comparison of FPKM converted TPM and count converted TPM is identical.

```{r}
rawcount <- read.table("easy_input.txt",
                        row.names = 1, header = TRUE, sep="\t")
count2tpm.mt <- read.table("output_count2tpm.txt", sep = "\t", 
                         row.names = 1)
count2tpm.mt[1:3,]
fpkm2tpm.mt <-  read.table("TCGA_FPKM2TPM.genes.txt", sep = " ",
                           row.names = 1)
fpkm2tpm.mt[1:3,]

# 比较原始计数和 count2tpm
# compare raw count and count2tpm
cor(x=rawcount$TCGA.DD.A11D.01A.11R.A131.07, y=count2tpm.mt$TCGA.DD.A11D.01A.11R.A131.07)
# 比较原始计数和 fpkm2tpm
# compare raw count with fpkm2tpm
cor(x=rawcount$TCGA.DD.A11D.01A.11R.A131.07,
    y=fpkm2tpm.mt$TCGA.DD.A11D.01A.11R.A131.07)
# 比较 count2tpm 和 fpkm2tpm
# compare  count2tpm and fpkm2tpm
cor(x=count2tpm.mt$TCGA.DD.A11D.01A.11R.A131.07,
    y=fpkm2tpm.mt$TCGA.DD.A11D.01A.11R.A131.07)
```

以第一列为例，raw count 和 count2tpm的相关系数是0.898，raw count 和fpkm2tpm的相关系数也是 0.898。而count2tpm和fpkm2tpm的相关系数是1, 证明了count2tpm的代码是有效的。


注2: biomaRt支持多物种, 通过更改参数biomart 和 host 来获取，植物的host跟动物的不同

Take the first column as an example, the correlation coefficient between raw count and count2tpm is 0.898, and the correlation coefficient between raw count and fpkm2tpm is also 0.898. The correlation coefficient between count2tpm and fpkm2tpm is 1, which proves that the code of count2tpm is valid.


Note 2: biomaRt supports multi-species, by changing the parameters biomart and host, the host of plants is different from that of animals.

```{r}
biomaRt::listMarts(host="plants.ensembl.org")
mart <- biomaRt::useMart(biomart = 'plants_mart',
                         host = "plants.ensembl.org")
genome <- biomaRt::listDatasets(mart)
```

注3:`biomaRt::getBM`由于要从国外服务器下载数据所以比较慢，可能还会断线, 当你遇到"The query to the BioMart webservice returned an invalid result: biomaRt expected a character string of length 1. Please report this to the mailing list."时不要慌张，反复重新运行这行命令就行

注4：如果无法通过`source("https://bioconductor.org/biocLite.R")`安装biocLite,那么就需要手动安装

把<https://bioconductor.org/biocLite.R>下载到本地并打开，在文件开头加入如下两行命令并保存。

Note 3:`biomaRt::getBM` is slower and may be disconnected because it has to download data from foreign server, when you encounter "The query to the BioMart webservice returned an invalid result: biomaRt expected a character string of length 1. Please report this to the mailing list.", don't panic, just re-run the command repeatedly.

Note 4: If you can't install biocLite via `source("https://bioconductor.org/biocLite.R")`, then you need to install it manually.

Download <https://bioconductor.org/biocLite.R> locally and open it, add the following two lines at the beginning of the file and save it.

```bash
#bioconductor

	CRANextra = "http://www.stats.ox.ac.uk/pub/RWin"))
#cran
```

然后使用 source调用本地biocLite.R文件安装bioclnstaller，再进行安装

Then use source to call the local biocLite.R file to install bioclnstaller, then install it.

```r
source('/path/to/biocLite.R')
BiocInstaller::biocLite()
```

```{r}
sessionInfo()
```
