---
title: "FigureYa222PCAgene"
params:
  author: "Zongcheng Li; Ying Ge, Yijing Chen"
output: html_document
---

**Author(s)**: `r params$author`  
**Date**: `r Sys.Date()`  


# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Requirement description

这个PCA很特别

This PCA is quite special.

![](example.png)

出自<https://linkinghub.elsevier.com/retrieve/pii/S0092867420300568>

图3. 卵母细胞亚型在逐步发育阶段的动态基因表达模式
(A) 基于主成分1（PC1）和主成分2（PC2）所呈现的基因表达模式，PCA图显示四种卵母细胞亚型。
(B) 沿PC1维度，对卵母细胞发育至关重要的代表性基因的相对基因表达模式。

From<https://linkinghub.elsevier.com/retrieve/pii/S0092867420300568>

Figure 3. Dynamic Gene-Expression Patterns of Oocyte Subtypes at Stepwise Developmental Stages
(A) PCA plot showing four oocyte subtypes based on gene-expression patterns exhibited by PC1 and PC2.
(B) Relative gene-expression patterns of representative genes essential for oocyte development along the PC1 dimension.

# 应用场景
# Application scenario

单细胞RNA-seq，从PCA结果就能看出基因表达模式跟细胞类型的关系。

原文：Principal component analysis (PCA) revealed that the four oocyte subtypes were distributed along the principal component 1 (PC1) dimension (Figure 3A; Table S2). 

Consistently, the relative expression levels of **genes known to be essential for follicular development also varied along PC1 axis** (Figures 3B and S3A). Genes promoting follicle development, such as ZP1, BMP15, and GDF9, were progressively **upregulated from oocyte subtype C1 to C4**. Meiotic M phase genes WEE2 and AURKA as well as DNA methyltransferase, DNMT1 and DNMT3A, were also gradually upregulated from subtype C1 to C4 (Figures 3B and S3A). 

Single-cell RNA-seq, the relationship between gene expression patterns and cell types can be discerned directly from the PCA results.

Original text: Principal component analysis (PCA) revealed that the four oocyte subtypes were distributed along the principal component 1 (PC1) dimension (Figure 3A; Table S2). 

Consistently, the relative expression levels of **genes known to be essential for follicular development also varied along PC1 axis** (Figures 3B and S3A). Genes promoting follicle development, such as ZP1, BMP15, and GDF9, were progressively **upregulated from oocyte subtype C1 to C4**. Meiotic M phase genes WEE2 and AURKA as well as DNA methyltransferase, DNMT1 and DNMT3A, were also gradually upregulated from subtype C1 to C4 (Figures 3B and S3A). 

# 环境设置
# Environment setting

```{r}
source("install_dependencies.R")

library(Seurat)
library(magrittr)
library(ggplot2)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 display English error messages
options(stringsAsFactors = FALSE) #禁止chr转成factor prohibit the conversion of chr to factor
```

# 输入文件
# Input file

sc.seurat.Rdata，单细胞RNA-seq预处理获得的文件，出自FigureYa206scHeatmap

怎样获得这个文件？拖到文末看“附：单细胞RNA-seq数据预处理”

sc.seurat.Rdata, the file obtained from single-cell RNA-seq preprocessing. It originates from FigureYa206scHeatmap.

How to obtain this file? Scroll to the end of the document and see "Appendix: Single-cell RNA-seq Data Preprocessing"

```{r}
(load("sc.seurat.Rdata"))
```

# PCA

## 原文采用的聚类方法
## The clustering method employed in the original text

例文的非监督聚类方法比较特殊，以SCENIC的分析结果作为输入（SCENIC的用法可参考FigureYa194pySCENIC），用modified ‘clustering-and-classification’ algorithm (Lake et al., 2016) 给细胞做聚类。clustering-and-classification也是个很有趣的半监督聚类算法。

The unsupervised clustering method used in the example text is quite unique, taking the analysis results from SCENIC as input (for the usage of SCENIC, you can refer to FigureYa194pySCENIC), and applying modified 'clustering-and-classification' algorithm (Lake et al., 2016) to cluster the cells. The clustering-and-classification is also an interesting semi-supervised clustering algorithm.

![](clustering.png)

## 我们用Seurat自带的无监督聚类
## We employ Seurat's built-in unsupervised clustering

```{r}
## 提取oocytes subtype的细胞的数据
# 从文章附件获得oocytes subtype的细胞ID（191 young oocytes, 227 old oocytes）
# 这里用到第一列cell ID，用于提取oocytes subtype的细胞
## Extract data for oocyte subtype cells
# Obtain cell IDs for oocyte subtypes from article attachments (191 young oocytes, 227 old oocytes)
# Utilise the first column cell ID to extract cells for oocyte subtypes
meta_oo <- readxl::read_excel(path = "1-s2.0-S0092867420300568-mmc2.xlsx", sheet = 2) 
sc_oo <- sc[,meta_oo$cell] # only 414 (not 418?! ) cells in the metadata 

## 自己的数据，就不涉及取子集，直接用sc就好，用下面这行代替上面两行：
## For your own data, there's no need to take a subset; you can directly use sc, replace the two lines above with the following:
# sc_oo <- sc

sc_oo_mnn <- batchelor::mnnCorrect(# mmnCorrect is no longer in scran mmnCorrect不再包含在scran中
  as.SingleCellExperiment(sc_oo),
  batch = sc_oo$individual,
  k = 4, 
  sigma = 0.1, cos.norm.in = T, cos.norm.out = T, 
  var.adj = T
)
prData <- SummarizedExperiment::assay(sc_oo_mnn)

# HVGs
#sc_oo %<>% FindVariableFeatures(selection.method = "mvp", mean.cutoff = c(1,8))
sc_oo %<>% FindVariableFeatures(selection.method = "vst")
VariableFeatures(sc_oo) %>% length()

# PCA
prres <- prcomp(prData[VariableFeatures(sc_oo), ] %>% t, scale. = F)
#prres <- prcomp(prData %>% t, scale. = F)
#prres$x <- sc_oo@reductions$pca@cell.embeddings

# unsupervised clustering based on mnn matrix
sc_oo %<>% ScaleData 
sc_oo@assays$RNA@scale.data <- prData[rownames(sc_oo@assays$RNA@scale.data), colnames(sc_oo@assays$RNA@scale.data)]
sc_oo %<>% RunPCA
ElbowPlot(sc_oo)

sc_oo %<>% FindNeighbors(dims = 1:15) %>% FindClusters(resolution = 0.6)

pca.pv <- sc_oo@reductions$pca
# 每个基因的主成分
# The principal component of each gene
# pca.pv@feature.loadings
dim(pca.pv@feature.loadings)
```

# 开始画图
# Start drawing

## 画Fig. 3A
## Draw Fig. 3A

```{r}
#计算坐标轴标签
#Calculate axis labels
pc1.pv <- paste0(round(pca.pv@stdev[1], digits = 3) * 100, "%")
pc2.pv <- paste0(round(pca.pv@stdev[2], digits = 3) * 100, "%")

# 修改seurat_clusters的名字
# Rename the seurat_clusters
sc_oo@meta.data$seurat_clusters <- factor(sc_oo@meta.data$seurat_clusters, 
                                          levels = c("0", "1", "2", "3"), 
                                          labels = c("C1", "C2", "C3", "C4"))

# 画图
# draw
PCAPlot(sc_oo, group.by = "seurat_clusters") +
  scale_color_manual(values = c("darkmagenta", "steelblue", "chartreuse3", "red")) +
  xlab(paste0("PC1 (", pc1.pv,")")) + 
  ylab(paste0("PC2 (", pc2.pv,")")) 

# 保存到文件
# Save to file
ggsave("PCA_A.pdf", width = 6, height = 5)
```

## 画Fig. 3B
## Draw Fig. 3B

用Seurat自带的无监督聚类获得的seurat_clusters画图

Plotting the seurat_clusters obtained using Seurat's built-in unsupervised clustering

```{r}
#ggplot() + geom_point(mapping = aes(prres$x[,1],  prres$x[,2], color = sc_oo$seurat_clusters))

markers <- c("ZP1", "WEE2", "DNMT1", "ATP6", "FIGLA", "SOX17")
reshape2::melt(cbind(sc_oo@meta.data[,"seurat_clusters", drop = F],
                     PC1 = -prres$x[,1],
                     GetAssayData(sc_oo)[markers,] %>% t
                     #SummarizedExperiment::assay(sc_oo_mnn)[markers,] %>% t
), 
id.vars = c("seurat_clusters", "PC1")) %>%
  ggplot(mapping = aes(PC1, value)) +
  ylab("Relative expression") +
  geom_point(mapping = aes(color = seurat_clusters)) +
  geom_smooth() +
  scale_color_manual(values = c("darkmagenta", "steelblue", "chartreuse3", "red")) +
  facet_wrap(~variable, scales = "free_y") +
  theme_classic() +
  theme(panel.border = element_rect(fill = NA),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
ggsave("PCA_B.pdf", width = 7, height = 4)
```

## 跟原文对比
## Compare with the original text

1-s2.0-S0092867420300568-mmc2.xlsx的第二列是细胞所在的cluster，是原文采用的聚类方法获得的cluster信息，用它跟我们做的“Seurat自带的无监督聚类”的结果对比。

没有完全重复出原文的效果。最可能的原因是mnn算法计算出的数据有差别，或者是作者选用的输入基因是特别的，原文方法中没有描述。

The second column of 1-s2.0-S0092867420300568-mmc2.xlsx indicates the cluster to which each cell belongs, representing the cluster information obtained by the clustering method used in the original text. This can be compared with the results of our "unsupervised clustering with Seurat's built-in method".

The results do not perfectly replicate the effects of the original text. The most likely reasons are differences in the data calculated by the MNN (Multimodal Neighborhood Network) algorithm or the use of specific input genes by the authors, which are not described in the original method.

```{r}
meta_oo <- readxl::read_excel(path = "1-s2.0-S0092867420300568-mmc2.xlsx", sheet = 2)
dim(meta_oo)
sc_oo$subtype <- meta_oo$cluster[match(colnames(sc_oo), meta_oo$cell)]

table(sc_oo$subtype, sc_oo$seurat_clusters)
```

用原文聚类获得的subtype画图

Plot the subtypes obtained from original text clustering

```{r}
#ggplot() +
  geom_point(mapping = aes(prres$x[,1],  prres$x[,2], color = sc_oo$subtype))

markers <- c("ZP1", "WEE2", "DNMT1", "ATP6", "FIGLA", "SOX17")
reshape2::melt(cbind(sc_oo@meta.data[,"subtype", drop = F],
        PC1 = -prres$x[,1],
        GetAssayData(sc_oo)[markers,] %>% t
        #SummarizedExperiment::assay(sc_oo_mnn)[markers,] %>% t
        ), 
  id.vars = c("subtype", "PC1")) %>%
  ggplot(mapping = aes(PC1, value)) +
  ylab("Relative expression") +
  geom_point(mapping = aes(color = subtype)) +
  geom_smooth() +
  scale_color_manual(values = c("darkmagenta", "steelblue", "chartreuse3", "red")) +
  facet_wrap(~variable, scales = "free_y") +
  theme_classic() +
  theme(panel.border = element_rect(fill = NA),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
ggsave("PCA_B_ori.pdf", width = 7, height = 4)
```

# 附：单细胞RNA-seq数据预处理
# Appendix: Single-cell RNA-seq Data Preprocessing

以下代码出自`FigureYa206scHeatmap`，会输出`sc.seurat.Rdata`，可作为以上代码的输入文件。

The following code is from `FigureYa206scHeatmap` and will output `sc.seurat.Rdata`, which can serve as the input file for the above code.

## 下载单细胞RNA-seq数据
## Download single-cell RNA-seq data

1) UMI count，从NCBI[GSE130664](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130664)下载：`GSE130664_merge_UMI_count.txt.gz`文件。

1) UMI count, download from NCBI[GSE130664](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130664): `GSE130664_merge_UMI_count.txt.gz` file.

![](download.png)

```{r download, eval=FALSE}
download.file("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE130664&format=file&file=GSE130664%5Fmerge%5FUMI%5Fcount%2Etxt%2Egz", 
              destfile = "GSE130664_merge_UMI_count.txt.gz")
```

2) metadata，从[例文的](https://doi.org/10.1016/j.cell.2020.01.009)Supplementary Tables获得：`1-s2.0-S0092867420300568-mmc1.xlsx`

2) Metadata, obtained from the Supplementary Tables of [the example paper](https://doi.org/10.1016/j.cell.2020.01.009): `1-s2.0-S0092867420300568-mmc1.xlsx`

## 读取数据
## Read data

```{r preprocessing, eval=FALSE}
umi <- read.table(file = gzfile("GSE130664_merge_UMI_count.txt.gz"), header = T, row.names = 1, sep = "\t")
qc <- readxl::read_excel("1-s2.0-S0092867420300568-mmc1.xlsx", sheet = 2)
meta <- readxl::read_excel("1-s2.0-S0092867420300568-mmc1.xlsx", 3) %>% 
  column_to_rownames("cell")
```

## 数据预处理
## Data preprocessing

参见方法部分：
定量和统计分析 -> 单细胞RNA-Seq数据处理。

See Methods:  
QUANTIFICATION AND STATISTICAL ANALYSIS -> Single-Cell RNA-Seq Data Processing

```{r, eval=FALSE}
# 细胞质量控制
# QC of Cells
cells <- qc %>% 
  filter(`Mapping rate` >= 0.2 &
           `Gene number` >= 700 &
           UMI >= 3000) %>%
  pull(Rename)

# seurat 对象
# seurat object
sc <- CreateSeuratObject(counts = umi[,cells], meta.data = meta)

# 表达量转换
# expression transformation
sc@assays$RNA@data <- sc@assays$RNA@counts %>% 
  apply(2, function(x){
    log2(10^5*x/sum(x)+1)
    })

# 去除其他细胞
# remove other cells
sc <- sc[,sc$cluster != "other"]

# 给cluster改名
# rename the clusters
sc$cluster_short <- factor(
  plyr::mapvalues(sc$cluster, 
                  c("Oocyte", "Natural killer T cell", "Macrophage",
                    "Granulosa cell", "Endothelial cell", 
                    "Smooth muscle cell", "Stromal cell"),
                  c("OO", "NKT", "M", "GC", "EC", "SMC", "SC")),
  levels = c("OO", "NKT", "M", "GC", "EC", "SMC", "SC"))

# 给cluster自定义颜色
# customise colours for the cluster
cluster_colors <- setNames(brewer.pal(7, "Set1"), levels(sc$cluster_short))

# 保存一下，便于停下来接着跑
# save the progress so that you can stop and resume later
save(sc, cluster_colors, file = "sc.seurat.Rdata")

# 还可以把表达矩阵输出到文件
# you can also output the expression matrix to a file
#write.csv(sc@assays$RNA@data, "easy_input_expr.csv", quote = F)
```

# Session Info

```{r}
sessionInfo()
```