---
title: "tSNE & Clustering"
params:
  author: "Jarning"
output: html_document
---

**Author(s)**: `r params$author`  
**Date**: `r Sys.Date()`  

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005




```{r}
source("install_dependencies.R")
knitr::opts_chunk$set(eval = FALSE, warning = FALSE, message = FALSE)
```

```{r, eval=TRUE}
oldpath <- .libPaths()[1]
newpath <- "/home/biostacs/R/backup/env_31723062"
```


```{r, eval=TRUE}
library(ggplot2)
library(Seurat, lib.loc = newpath)
```

导入第二部分的结果
```{r}
if (!exists("seu.CCA")) {
  seu.CCA <- readRDS("tmp/s2-seu_CCA.rds")
}

seu.CCA <- RunTSNE(seu.CCA, reduction.use = "cca.aligned", dims.use = 1:20)
```

```{r, include=FALSE, eval=TRUE}
seu.CCA <- readRDS("tmp/s3.2-seu_CCA_clustering.rds")
```

查看tSNE降维图，检查批次效应矫正的效果
```{r, fig.width=12, fig.height=8, eval=TRUE}
tsne.emb <- GetDimReduction(seu.CCA, reduction.type = "tsne", slot = "cell.embeddings")
tsne.emb <- as.data.frame(tsne.emb)
tsne.emb$sample_id <- as.character(seu.CCA@meta.data[rownames(tsne.emb),]$sample_id)
ggplot(data = tsne.emb, aes(x=tSNE_1, y=tSNE_2, color=sample_id)) + 
  geom_point(size = .1) +
  facet_wrap(~sample_id, ncol = 3) + 
  theme_bw() + 
  theme(legend.position = "none",
        strip.text = element_text(size = 15, face = "bold"))
```

> Clusters were identifed using the function “FindClusters” from Seurat using default parameters.

聚类
```{r, eval=TRUE}
seu.CCA <- FindClusters(seu.CCA, reduction.type = "cca.aligned", dims.use = 1:20, save.SNN = TRUE)
```


对聚类结果进行评价<br>

> The robustness of the clusters was calculated using the function “AssessNodes” from Seurat. For each cluster, a phylogenetic tree based on the distance matrix in gene expression space was computed. Next, Seurat computed an out-of-bag error for a random forest classifer trained on each internal node split of the tree.

构建phylogenetic tree 
```{r, eval=TRUE}
seu.CCA <- BuildClusterTree(seu.CCA, reorder.numeric = TRUE, do.reorder = TRUE, do.plot = FALSE)
```

计算out-of-bag error，文章认为oobe<0.1的节点分类是可靠的。
```{r, eval=TRUE}
oob <- AssessNodes(seu.CCA)
oob
```

可以看到节点37的oobe=0.138>0.1，我们可以将节点37的两个分类合并，从phylogenetic tree上看，就是cluster2和cluster3。

```{r, eval=TRUE}
PlotClusterTree(seu.CCA)
```


```{r, fig.width=12, fig.height=8, eval=TRUE}
tsne.emb$ident <- factor(as.character(seu.CCA@ident), levels = levels(seu.CCA@ident))

ggplot(data = tsne.emb, aes(x=tSNE_1, y=tSNE_2, color=ident)) + 
  geom_point(size = .1) +
  facet_wrap(~sample_id, ncol = 3) + 
  guides(colour = guide_legend(title = "Cluster", 
                               title.hjust=0.5, 
                               keyheight = 1.5, 
                               override.aes = list(size=6))) + 
  theme_bw() + 
  theme(strip.text = element_text(size = 15, face = "bold"))
```

查看cluster2和cluster3
```{r, fig.width=5, fig.height=4, eval=TRUE}
tsne.emb$is_2_3 <- tsne.emb$ident %in% c(2,3)

ggplot(data = tsne.emb, aes(x=tSNE_1, y=tSNE_2, color=is_2_3)) + 
  geom_point(size = .1) +
  scale_color_manual(values = c("grey","red")) + 
  guides(colour = guide_legend(title = "Cluster 2&3",
                               title.hjust=0.5, 
                               keyheight = 1.5, 
                               override.aes = list(size=6))) + 
  theme_bw() + 
  theme(strip.text = element_text(size = 15, face = "bold"))
```

合并节点37，重新构建phylogenetic tree
```{r, eval=TRUE}
seu.CCA <- MergeNode(seu.CCA, node.use = 37, rebuild.tree = TRUE, 
                     reorder.numeric = TRUE, do.reorder = TRUE, do.plot = FALSE)
plotClusterTree(seu.CCA)
```

重新计算out-of-bag error
```{r, eval=TRUE}
oob <- AssessNodes(seu.CCA)
oob
```

这下所有节点的oobe<0.1，认为此时的聚类是可靠的(我们得到了18个cluster,文章中给出了16个cluster)

```{r, fig.width=12, fig.height=8, eval=TRUE}
tsne.emb$ident <- factor(as.character(seu.CCA@ident), levels = levels(seu.CCA@ident))

ggplot(data = tsne.emb, aes(x=tSNE_1, y=tSNE_2, color=ident)) + 
  geom_point(size = .1) +
  facet_wrap(~sample_id, ncol = 3) + 
  guides(colour = guide_legend(title = "Cluster", 
                               title.hjust=0.5, 
                               keyheight = 1.5, 
                               override.aes = list(size=6))) + 
  theme_bw() + 
  theme(strip.text = element_text(size = 15, face = "bold"))
```

保存计算结果
```{r}
saveRDS(oob, "tmp/s3.1-clustering_robustness.rds")
saveRDS(seu.CCA, "tmp/s3.2-seu_CCA_clustering.rds")
```
