---
title: "FigureYa213customizeHeatmap"
params:
  author: "Xiaofan Lu"
output: html_document
reviewers: Ying Ge,Hui Huang
---

**Author(s)**: `r params$author`  
**Date**: `r Sys.Date()`  

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

Figure6 的绘制方法以及Cmap score的计算，图里其他的内容如何获取可以无需讲解。

# Requirement Description

The drawing method of Figure 6 and the calculation of Cmap score, and how to obtain the rest of the content in the figure do not need to be explained.

![](example.png)

出自<https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbaa164/5891146>
from<https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbaa164/5891146>

Figure 6. Identification of most promising therapeutic agents for high PPS score patients according to the **evidence from multiple sources** (see also Table S15 available online at https://academic.oup.com/bib). **Six CTRP-derived agents** and **six PRISM-derived agents** were shown on the **left** and **right** of the diagram, respectively.

FigureYa131CMap也用到了CMap，并对感兴趣的小分子进行MoA药物作用机制分析，可相互参考。

CMap is also used in FigureYa131CMap, and MoA drug mechanism of action analysis is performed on the small molecule of interest, which can be cross-referenced.

# 应用场景

例文采用了多个来源的数据相互佐证，来说明筛选出的化合物/药物的潜在价值。作者设计了这个图，巧妙地把证据一、二、三和四同时展示出来，同时还能看到overlap。

- 证据一和二：上文用PRISM和CTRP发现了12个 candidate compounds identified showed a higher drug sensitivity in PPS score-high patients（可参考FigureYa212drugTarget）
- 证据三：We first used the **CMap analysis** to find compounds of which gene expression patterns were oppositional to the HCC-specific expression patterns (i.e. gene expression increased in tumor tissues but decreased by treatment of certain compounds).
- 证据四：Secondly, **fold-change differences of the expression levels (including mRNA- and protein-level) of candidates’ drug targets** between tumor and normal tissue were calculated, and a higher fold change value indicated a greater potential of candidate agent for HCC treatment. 
- 证据五：Thirdly, a comprehensive **literature search** was performed in PubMed (https://www.ncbi.nlm.nih. gov/pubmed/) to find out the experimental and clinical evidence of candidate compounds in treating HCC. 

# Application Scenarios

Data from multiple sources are used to corroborate each other to illustrate the potential value of the screened compounds/drugs. The author has designed this diagram to cleverly display evidence one, two, three, and four at the same time, while also seeing the overlap.

- Evidence 1 & 2: 12 candidate compounds identified showed a higher drug sensitivity in PPS score-high patients using PRISM and CTRP above (see FigureYa212drugTarget)
- Evidence 3: We first used the **CMap analysis** to find compounds of which gene expression patterns were oppositional to the HCC-specific expression patterns (i.e. gene expression increased in tumor tissues but decreased by treatment of certain compounds).
- Evidence 4: Secondly, **fold-change differences of the expression levels (including mRNA- and protein-level) of candidates' drug targets** between tumor and normal tissue were calculated, and a higher fold change value indicated a greater potential of candidate agent for HCC treatment.
- Evidence 5: Thirdly, a comprehensive **literature search** was performed in PubMed (https://www.ncbi.nlm.nih. gov/pubmed/) to find out the experimental and clinical evidence of candidate compounds in treating HCC.

下面将带你实现CMap analysis及画图。以左图CTRP vs. CMap为例，右图也是一样的。

> 使用本代码或分析思路请引用：

The following will take you through the implementation of CMap analysis and drawing. Take CTRP vs. CMap on the left as an example, and the same is true for the right.

> To use this code or analysis idea, please quote:

Chen Yang, Xiaowen Huang, Yan Li, Junfei Chen, Yuanyuan Lv, Shixue Dai, Prognosis and personalized treatment prediction in TP53-mutant hepatocellular carcinoma: an in silico strategy towards precision oncology, Briefings in Bioinformatics, bbaa164, https://doi.org/10.1093/bib/bbaa164

# 环境设置

画图用到了github最新版ComplexHeatmap里的pheatmap，两种方式都可以用上它，任选其一。

方式一：安装github版的ComplexHeatmap（Xiaofan Lu亲测有效）

# Environment settings

```{r}
source("install_dependencies.R")
```

加载新ComplexHeatmap包里的pheatmap函数
Load the pheatmap function in the new ComplexHeatmap package

```{r}
source("pheatmap_translate.R") # 来自github版ComplexHeatmap # From github version of ComplexHeatmap
```

```{r}
library(devtools)
library(ComplexHeatmap) # 用于绘制热图 # Used to draw heatmaps
library(circlize) # 用于配色 # For color matching
library(tidyverse) # 用于读取MAF文件 # Used to read MAF files
library(limma) # 用于差异表达 # For differential expression

Sys.setenv(LANGUAGE = "en") # 显示英文报错信息  # error messages are displayed in English
options(stringsAsFactors = FALSE) # 禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入文件

如果你只是想画图，就直接跳到“开始画图”。

LIHC.TPM.txt，基因表达矩阵，用于计算差异表达倍数。与FigureYa212drugTarget的相同。

data_mutations_mskcc.txt，maf突变数据，用于提取亚型样本，此处以TP53突变为例。与FigureYa212drugTarget相同。

# Input files

If you just want to draw, just skip to "Start drawing".

LIHC. TPM.txt, gene expression matrix, which is used to calculate differential expression folds. Same as FigureYa212drugTarget.

data_mutations_mskcc.txt, maf mutation data were used to extract isoform samples, and the TP53 mutation was used as an example. Same as FigureYa212drugTarget.

```{r}
# 1.读取肝癌TPM表达谱
# 1.Read the TPM expression profile of hepatocellular carcinoma

expr <- read.table("LIHC.TPM.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
normsam <- colnames(expr[,which(substr(colnames(expr),11,12) == "11")])
tumosam <- colnames(expr[,which(substr(colnames(expr),11,12) == "01")])

# 2.读取maf突变文件(与cBioPortal下载)并更新表达谱
# 2. Read the maf mutation file (download with cBioPortal) and update the expression profile

maf <- read_tsv("data_mutations_mskcc.txt", comment = "#")
maf$Tumor_Sample_Barcode <- paste0("LIHC",substr(maf$Tumor_Sample_Barcode,8,15))
tumosam <- intersect(tumosam,unique(maf$Tumor_Sample_Barcode)) # 取出交集的肿瘤样本 # Remove the intersecting tumor samples
maf <- maf[which(maf$Tumor_Sample_Barcode %in% tumosam),]
expr <- expr[,c(tumosam,normsam)]

# 3.提取TP53突变信息并创建样本注释
# 3. Extract TP53 mutation information and create sample annotations

tp53 <- c()
for (i in tumosam) {
  tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
  if(is.element("TP53", tmp$Hugo_Symbol)) { # 如果存在TP53 # If TP53 is present
    tp53 <- c(tp53,1)  #记录1 # Record 1
  } else {
    tp53 <- c(tp53,0) # 否则记录0 # Otherwise record 0
  }
}
names(tp53) <- tumosam
tp53.mutsam <- names(tp53[which(tp53 == 1)]) # 取出有TP53突变的患者 # Extraction of patients with TP53 mutations
```

# 差异表达分析(肿瘤vs正常)
# Differential expression analysis (tumor vs normal)

```{r}
pd <- data.frame(Samples = c(tp53.mutsam,normsam),
                 Group = rep(c("tumor","normal"),c(length(tp53.mutsam),length(normsam))),
                 stringsAsFactors = FALSE)
design <-model.matrix(~ -1 + factor(pd$Group, levels = c("tumor","normal")))
colnames(design) <- c("tumor","normal")
gset <- log2(expr[,pd$Samples] + 1)
fit <- limma::lmFit(gset, design = design);
contrastsMatrix <- limma::makeContrasts(tumor - normal, levels = c("tumor", "normal"))
fit2 <- limma::contrasts.fit(fit, contrasts = contrastsMatrix)
fit2 <- limma::eBayes(fit2, 0.01)
resData <- limma::topTable(fit2, adjust = "fdr", sort.by = "B", number = 100000)
resData <- as.data.frame(subset(resData, select=c("logFC","t","B","P.Value","adj.P.Val")))
resData$id <- rownames(resData)
colnames(resData) <- c("log2fc","t","B","pvalue","padj","id")
resData$fc <- 2^resData$log2fc
resData <- resData[order(resData$padj),c("id","fc","log2fc","pvalue","padj")]

# 保存到文件
# Save to file

write.table(resData, file = "output_degs.txt", row.names = FALSE, sep = "\t", quote = FALSE)
```

# CMap 分析

## 生成CMap的输入文件

根据原文补充材料的描述：300 genes with the most significant fold changes (150 up-regulated genes and 150 down-regulated genes) were submitted to the CMap website <https://clue.io/query>

# CMap analysis

## Generate input files for CMap

According to the original supplemental description: 300 genes with the most significant fold changes (150 up-regulated genes and 150 down-regulated genes) were submitted to the CMap website <https://clue.io/query>

```{r}
ngene <- 150
degs <- na.omit(resData[order(resData$log2fc,decreasing = T),])
updegs <- rownames(degs)[1:ngene]
dndegs <- rownames(degs)[(nrow(degs)-ngene + 1):nrow(degs)]
cmap.input <- data.frame(up = updegs,
                         dn = dndegs,
                         stringsAsFactors = F)
write.table(cmap.input,"CMap_input.txt",sep = "\t",row.names = F,col.names = T,quote = F)
```

## CMap网站使用

输入CMap_input.txt，经过以下方法获得CMap_export.txt文件，用于下一步画图。

## CMap Website Usage

Enter the CMap_input.txt and go through the following method to obtain a CMap_export.txt file for the next step of drawing.

![](CMap_1.jpg)

![](CMap_2.jpg)

![](CMap_3.jpg)

# 开始画图

图分为左中右三部分，左侧区块相当于annotation，中间为基因变化倍数，右侧为CMap score。

把annotation、log2FC、以及前面计算得到的CMap score填进去。

这里直接把数值写入矩阵matrix，实际使用时可通过读入存储相应数值的文件来赋值。

# Start drawing

The figure is divided into three parts: left, middle and right, the left block is equivalent to the annotation, the middle is the gene change fold, and the right side is the CMap score.

Fill in the annotation, log2FC, and the CMap score calculated earlier.

Here, the values are directly written into the matrix, and the values can be assigned by reading into the file that stores the corresponding values.

```{r}
cmap.res <- read.delim("CMap_export.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
drug <- c("BI-2536","Leptomycin B","methotrexate","narciclasine","SR-II-138A","vincristine")
selres <- cmap.res[which(cmap.res$Name %in% intersect(cmap.res$Name,drug)),]
print(selres) 
```

下面根据这个结果构建matrix，用于绘制热图，分别准备：

- 伪数据[用于映射颜色]，dt
- 展示标签，lb
- 标签对应颜色，cl

Let's build a matrix based on this result, which will be used to draw a heat map, and prepare them separately:

- Pseudo-data [for mapping colors], dt
- Display labels, lb
- The label corresponds to the color, cl

```{r}
## 左侧区块
# 背景颜色
## Block on the left
# Background color
dt1 <- matrix(c(0,1, # 1为深色，0为浅色 # 1 is dark and 0 is light
                0,1,
                1,1,
                0,0,
                0,0,
                1,1),
              ncol = 2,
              byrow = T,
              dimnames = list(c("BI-2536","Leptomycin B","Methotrexate","Narciclasine","SR-II-138A","Vincristine"),
                              c("Clinical status","Experimental evidence")))

# 文字标签
# Text labels
lb1 <- matrix(c("Phase2","Present",
                "Phase1","Present",
                "Launched","Present",
                "Preclinical","Absent",
                "Preclinical","Absent",
                "Launched","Present"),
              ncol = 2,
              byrow = T)

# 文字颜色
# Text color
cl1 <- matrix(c("black","white",
                "black","white",
                "white","white",
                "black","black",
                "black","black",
                "white","white"),
              ncol = 2,
              byrow = T)

# 画图
# 如果这里报错，请返回开头，看“环境设置”
# Drawing
# If an error is reported here, please go back to the beginning and look at "Environment Settings"
hm1 <- pheatmap(mat = dt1,
                color = c("#EFFAE8","#BAE5BC"),
                cluster_cols = F,
                cluster_rows = F,
                show_rownames = T,
                show_colnames = T,
                display_numbers = lb1,
                number_color = cl1,
                fontsize_number = 11,
                border_color = "white",
                cellwidth = 50,
                cellheight = 50,
                legend = F)


## 中间区块
# 背景颜色
## Intermediate block
# Background color
dt2 <- matrix(c(1,0,
                1,0,
                0,0,
                1,0,
                0,0,
                1,0),
              ncol = 2,
              byrow = T,
              dimnames = list(c("BI-2536","Leptomycin B","Methotrexate","Narciclasine","SR-II-138A","Vincristine"),
                              c("Log2FC of mRNA expression","Log2FC of protein expression")))

# 文字标签
# Text labels
lb2 <- matrix(c("1.23","0.42",
                "1.21","0.43",
                "0.89","-0.15",
                "1.07","0.24",
                "0.59","0.06",
                "1.54","-0.01"),
              ncol = 2,
              byrow = T)
# 文字颜色
# Text color
cl2 <- matrix(c("white","black",
                "white","black",
                "black","black",
                "white","black",
                "black","black",
                "white","black"),
              ncol = 2,
              byrow = T)
# 画图
# Drawing
hm2 <- pheatmap(mat = dt2,
                color = c("#B7E4ED","#019AC9"),
                cluster_cols = F,
                cluster_rows = F,
                show_rownames = T,
                show_colnames = T,
                display_numbers = lb2,
                number_color = cl2,
                fontsize_number = 11,
                border_color = "white",
                cellwidth = 50,
                cellheight = 50,
                legend = F)

## 右侧区块
# 背景颜色
## Right block
# Background color
dt3 <- matrix(c(0, # 0为中间色
                -1, # -1为最浅色
                0,
                1, # 1为深色
                -1,
                0),
              ncol = 1,
              byrow = T,
              dimnames = list(c("BI-2536","Leptomycin B","Methotrexate","Narciclasine","SR-II-138A","Vincristine"),
                              c("CMap score")))

# 文字标签
# 根据selres的结果填写
# 因为有两个药物没有CMap结果，所以是NA
# Text labels
# Fill in the form according to the result of SERes
# Because there are two drugs that do not have CMap results, it is NA
lb3 <- matrix(c("-83.13",
                "NA",
                "-49.61",
                "-91.23",
                "NA",
                "-34.30"),
              ncol = 1,
              byrow = T)

# 文字颜色
# Text color
cl3 <- matrix(c("black",
                "black",
                "black",
                "white",
                "black",
                "black"),
              ncol = 1,
              byrow = T)

# 画图
# Drawing
hm3 <- pheatmap(mat = dt3,
                color = c("white","#FDCEB9","#F26E5F"),
                cluster_cols = F,
                cluster_rows = F,
                show_rownames = T,
                show_colnames = T,
                display_numbers = lb3,
                number_color = cl3,
                fontsize_number = 11,
                border_color = "white",
                cellwidth = 50,
                cellheight = 50,
                legend = F)

## 水平合并热图
## Merge heatmaps horizontally
hm <- hm1 + hm2 + hm3
draw(hm) 
dev.copy2pdf(file = "heatmap.pdf",width = 8,height = 8)
```
作者表示整张图片均由热图拼接AI绘制，这里输出的pdf文件是矢量图，图片剩余细节以及图例请使用AI绘制。

The author said that the whole picture is drawn by AI after the heat map is stitched, and the pdf file output here is a vector image, and the remaining details and legends of the picture should be drawn by AI.
# Session Info

```{r}
sessionInfo()
```