---
title: "FigureYa249Regulon"
params:
  author: "Xiaofan Lu"
output: html_document
reviewers: Ying Ge,Hui Huang
---

**Author(s)**: `r params$author`  
**Date**: `r Sys.Date()`  

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005

If you use ComplexHeatmap in published research, please cite:

Zuguang Gu, et al., Complex heatmaps reveal patterns and correlations in multidimensional genomic data, Bioinformatics, 2016.

Zuguang Gu. Complex Heatmap Visualization, iMeta, 2022.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

Figure2a中regulon的计算方法（看起来上半部分是癌种特异性的，下半部分是通用的，随便哪个都行）。

# Requirement Description

The calculation method of regulon in Figure 2a (it seems that the top half is cancer-specific, the bottom half is generic, whichever is fine).

![](example.png)

出自<https://www.biorxiv.org/content/10.1101/2021.05.30.446369v1.full>
from<https://www.biorxiv.org/content/10.1101/2021.05.30.446369v1.full>

Figure 2. Molecular landscape of four MIBC iCSs.
a) Heatmap showing **regulon activity profiles** for 
23 transcription factors (top panel，即下文复现的图), 
potential regulators associated with chromatin remodelling (middle panel), 
and 296 unique differentially methylated promoters derived from each iCS vs. adjacent normal samples (bottom panel).

# 应用场景

计算转录调控网络regulon的活性（注意：并非转录因子活性），画图展示各亚型中regulon的活性规律。

让你的表达谱分析逐层递进，可参考例文的做法：

Figure 2a热图分上中下三部分，把自己的表达数据跟转录因子、chromatin remodelling regulator、甲基化分析建立联系，探讨转录调控机制。看作者怎么描述结果的：

# Application Scenarios

The activity of the transcriptional regulatory network regulon was calculated (note: it is not transcription factor activity), and the activity of regulon in each isoform was plotted.

To make your expression profile analysis progressive, you can refer to the example method:

Figure 2a is divided into three parts: upper, middle and lower, and links your expression data with transcription factors, chromatin remodelling regulators, and methylation analysis to explore transcriptional regulatory mechanisms. See what the author has to say about the results:

To further explore transcriptomic differences, we analysed regulons for **23 MIBC-specific TFs** and potential **regulators relevant to cancerous chromatin remodelling** [5, 30], leading to a strong confirmation of the biological pertinency of the four-classification because the regulon activity was tightly associated with iCSs (Figure 2a)... 
Regulon activity profiles that were associated with cancerous **chromatin remodelling** highlighted other possible differential regulatory patterns among four iCSs, indicating that epigenetically driven transcriptional networks might be important differentiators of these molecular subtypes (Figure 2a). 
The potential epigenetic differences among the subtypes might be further supported by **differential methylation analysis**, which demonstrated that iLS3 (265 vs. 45 in iLS2) and iBS4 (191 vs. 26 in iBS1) had more hypermethylated promoters (296 unique loci) than the 21 adjacent normal bladder samples had (Figure 2a).

文中很多图我们都实现过，例如：

- Figure 1b的画法可参考FigureYa196PanPie
- Figure 2b的画法可参考FigureYa248MutLandscape
- Figure 3de可产考FigureYa25sankey和FigureYa125Fishertest
- Figure 4f可参考FigureYa106immunotherapy
- Figure 5b可参考FigureYa12box，f可参考FigureYa162boxViolin
- Figure 6可参考FigureYa35batch_bestSeparation，FigureYa144DiagHeatmap或<https://mp.weixin.qq.com/s/34WRZRBVPHUNRLlzNH2nzw>

Many of the pictures in the article have been crowdsourced, such as:

- For more information about Figure 1b, see FigureYa196PanPie
- For Figure 2b, see FigureYa248MutLandscape
- Figure 3de see FigureYa25sankey and FigureYa125Fishertest
- Figure 4f can refer to Figure Ya106 immunotherapy
- Figure 5b can refer to FigureYa12box, and f can refer to FigureYa162boxViolin
- For Figure 6, see FigureYa35batch_bestSeparation, FigureYa144DiagHeatmap, or <https://mp.weixin.qq.com/s/34WRZRBVPHUNRLlzNH2nzw>

# 环境设置

# Environment settings

```{r}
source("install_dependencies.R")

library(RTN)
library(snow)
library(ComplexHeatmap)
library(ClassDiscovery)
library(RColorBrewer)
library(gplots)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

自定义函数
Custom functions

```{r}
standarize.fun <- function(indata=NULL, halfwidth=NULL, centerFlag=T, scaleFlag=T) {  
  outdata=t(scale(t(indata), center=centerFlag, scale=scaleFlag))
  if (!is.null(halfwidth)) {
    outdata[outdata>halfwidth]=halfwidth
    outdata[outdata<(-halfwidth)]= -halfwidth
  }
  return(outdata)
}
```

# 输入文件

easyinput_tcga_tpm.txt，表达矩阵

easyinput_tcga_phenotype.txt，每个样本对应的亚型信息，这里有四个亚型。

easyinput_regulon.txt，运行regulon分析只需要基因的名字。

- 热图上中下三部分各对应一类基因（转录因子、chromatin remodelling regulator、甲基化）。
- 这里以最上面的23个转录因子为例，基因名出自<https://www.sciencedirect.com/science/article/pii/S0092867417310565?via%3Dihub>。
- 对23个转录因子的描述原文：a total of 23 ‘regulator’ genes that were associated induced/repressed targets: the steroid hormone receptors ESR1/2, AR and PGR; the nuclear receptors PPARG, three RARs (A/B/G), and three RXRs (A/B/G); the receptor tyrosine kinases ERBB2/3 and FGFR1/3; and the transcription factors FOXA1, FOXM1, GATA3/6, HIF1A, KLF4 and STAT3 and TP63
- **对于自己的数据，可以自己挖掘或从文献中获取感兴趣的组织、疾病甚至发育时期特异的调控因子。**

# Input files

easyinput_tcga_tpm.txt, the expression matrix。

easyinput_tcga_phenotype.txt, the subtype information corresponding to each sample, there are four subtypes here.

easyinput_regulon.txt, only the name of the gene is required to run the regulon analysis.

- The upper, middle, and lower parts of the heatmap correspond to a class of genes (transcription factor, chromatin remodelling regulator, methylation).
- Here are the top 23 transcription factors as an example, and the gene names are from <https://www.sciencedirect.com/science/article/pii/S0092867417310565?via=ihub>。
- a total of 23 'regulator' genes that were associated induced/repressed targets: the steroid hormone receptors ESR1/2, AR and PGR; the nuclear receptors PPARG, three RARs (A/B/G), and three RXRs (A/B/G); the receptor tyrosine kinases ERBB2/3 and FGFR1/3; and the transcription factors FOXA1, FOXM1, GATA3/6, HIF1A, KLF4 and STAT3 and TP63
- For your own data, you can dig your own or obtain specific regulators of interest in tissues, diseases, and even developmental periods from the literature. **

```{r}
# 加载基因表达以及样本数值信息
# Load gene expression and sample value information
tcgaBLCA <- read.table("easyinput_tcga_tpm.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
pheno <- read.table("easyinput_tcga_phenotype.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)

# 加载MIBC特异性的调控子
# Load MIBC-specific regulators
tfs <- read.table("easyinput_regulon.txt",header = T)
```

# 计算regulon活性 - Regulon analysis

bootstrap计算reference regulatory network运行时间较长。

为方便反复运行调试代码，把中间数据保存到了文件。可跳过这步，进入“开始画图”。

# Calculate regulon activity - Regulon analysis

Bootstrap calculates that the reference regulatory network takes a long time to run.

In order to facilitate the repeated running of the debugging code, the intermediate data is saved to a file. You can skip this step and go to Start Drawing.

```{r}
# 取共有的基因名
# Take the common gene name
regulatoryElements <- intersect(tfs$regulon, rownames(tcgaBLCA))

# 运行TNI构建程序
# Run the TNI builder
# we used the R package “RTN” to reconstruct transcriptional regulatory networks (regulons)
rtni_tcgaBLCA <- tni.constructor(expData = as.matrix(log2(tcgaBLCA + 1)), # 样图计算时候没有取对数,  # There is no logarithm when calculating the sample,
                                 regulatoryElements = regulatoryElements)

# 通过置换以及bootstrap计算reference regulatory network.
# Reference regulatory network is calculated by substitution and bootstrap.
# mutual information analysis and Spearman rank-order correlation deduced the possible associations between a regulator and all potential target from the transcriptome expression profile, and permutation analysis was utilized to erase associations with an FDR > 0.00001. Bootstrapping strategy removed unstable associations through one thousand times of resampling with consensus bootstrap greater than 95%. 
# 这里量力而行设置多核，或者直接单核运算
# Set up multi-core or single-core operation as you can
options(cluster=snow::makeCluster(spec = 4, "SOCK")) # 打开4核并行计算（不确定是不是4核，不过我windows只用4，服务器我开12）# Turn on 4-core parallel computing (I'm not sure if it's 4 cores, but I only use 4 for windows, and I open 12 for servers)
rtni_tcgaBLCA <- tni.permutation(rtni_tcgaBLCA, pValueCutoff = 1e-5, nPermutations = 1000)
rtni_tcgaBLCA <- tni.bootstrap(rtni_tcgaBLCA, nBootstraps = 1000)
stopCluster(getOption("cluster")) # 关闭并行计算 # Turn off parallel computing

# 计算DPI-filtered regulatory network
# Calculate the DPI-filtered regulatory network
# Data processing inequality filtering eliminated the weakest associations in triangles of two regulators and common targets
rtni_tcgaBLCA <- tni.dpi.filter(rtni_tcgaBLCA, eps = 0, sizeThreshold = TRUE, minRegulonSize = 15)

# 保存TNI对象以便后续分析
# Save the TNI object for later analysis
save(rtni_tcgaBLCA, file="rtni_tcgaBLCA.RData")

# load("rtni_tcgaBLCA.RData")
# 计算每个样本的regulon活性
# Calculate the regulon activity for each sample
# Individual regulon activity was estimated by two-sided GSEA
rtnigsea_tcgaBLCA <- tni.gsea2(rtni_tcgaBLCA, regulatoryElements = regulatoryElements)
MIBC_regact <- tni.get(rtnigsea_tcgaBLCA, what = "regulonActivity")

# 保存活性对象
# Save the active object
save(MIBC_regact,file = "MIBC_regact.RData") 
```

# 开始画图
# Start drawing

```{r}
# 加载活性对象
# Load the active object
(load("MIBC_regact.RData"))

# 设置颜色
# Set the color
clust.col <- c("#DD492E","#40548A","#32A087","#EC7D21")
blue <- "#5bc0eb"
gold <- "#ECE700"

plotdata <- standarize.fun(t(MIBC_regact$differential),halfwidth = 1.5) # 标准化regulon的活性 # Normalize the activity of regulon 
annCol.tcga <- pheno[order(pheno$CMOIC),,drop = F] # 构建样本注释信息，并对亚型进行排序 # Construct sample annotation information and sort subtypes
annColors.tcga <- list()
annColors.tcga[["CMOIC"]] <- c("CS1" = clust.col[1],
                               "CS2" = clust.col[2],
                               "CS3" = clust.col[3],
                               "CS4" = clust.col[4])
hcg <- hclust(distanceMatrix(as.matrix(MIBC_regact$differential[rownames(annCol.tcga),]), "euclidean"), "ward.D")
hm <- pheatmap(plotdata[hcg$order,rownames(annCol.tcga)],
               border_color = NA, # 热图单元格无边框 # Heatmap cells have no borders
               color = colorpanel(64,low=blue,mid = "black",high=gold),
               cluster_rows = F, # 行不聚类 # Rows are not clustered
               cluster_cols = F, # 列聚类 # Column clustering
               show_rownames = T, # 显示行名 # Show the rowname
               show_colnames = F, # 不显示列名 # Column names are not displayed
               gaps_col = cumsum(table(annCol.tcga$CMOIC))[1:3], # 亚型分割 # Subtype segmentation
               cellwidth = 0.8, # 固定单元格宽度 # Fixed cell width
               cellheight = 10, # 固定单元格高度 # Fixed cell height
               name = "MIBC Regulon", # 图例名字 # Legend name
               annotation_col = annCol.tcga[,"CMOIC",drop = F], # 样本注释 # Sample annotations
               annotation_colors = annColors.tcga["CMOIC"]) # 样本注释的对应颜色 # The color of the sample annotation
 
pdf("regulon heatmap.pdf", width = 8,height = 6)
draw(hm) # 输出热图 # Output heatmap
invisible(dev.off())
```

# Session Info

```{r}
sessionInfo()
```
